presto-main/src/main/java/com/facebook/presto/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -199,7 +199,20 @@ else if (pendingSplits.isEmpty()) {
                     SplitBatch nextSplits = getFutureValue(scheduleGroup.nextSplitBatchFuture);
                     scheduleGroup.nextSplitBatchFuture = null;
                     pendingSplits.addAll(nextSplits.getSplits());
-                    if (nextSplits.isLastBatch() && scheduleGroup.state == ScheduleGroupState.DISCOVERING_SPLITS) {
+                    if (nextSplits.isLastBatch()) {
+                        if (scheduleGroup.state == ScheduleGroupState.INITIALIZED && pendingSplits.isEmpty()) {
+                            // Add an empty split in case no splits have been produced for the source.
+                            // For source operators, they never take input, but they may produce output.
+                            // This is well handled by Presto execution engine.
+                            // However, there are certain non-scan operators that may produce output without any input,
+                            // for example, an AggregationOpreator (not HashAggregationOperator).
```

Comment:
The "not HashAggregationOperator" part is inaccurate. It can if it's processing grouping sets and the global grouping set is present.